### DISCUSSION TITLE
Request: Default sort supporting more than one column in the Table Builder

### DISCUSSION THREAD

**Discussion by maccarib** [filamentphp/filament#6095] [Ideas]
In the Table Builder you have the getDefaultTableSortColumn method available to achieve a default sorting by a given column.
In a admin resource table you can use defaultSort to sort by several columns.

Contrary to the resource table it is not possible in the Table Builder to achieve a default sort by two or more columns. Of course it is possible to put that sorting using orderBy in the table query. But then the interactive sorting won't work any longer as the table query orderBy always comes first and wins.

It would be helpful to allow for an array of sort columns in an equivalent method such as getDefaultTableSortColumns.

**Comment by danharrin** [comment]
Duplicate of other thread/s, and implemented in v3

**Reply by maccarib** [reply]
OK, thanks for the feedback. And sorry for the duplicate thread. I've searched for similar threads, but according to my understandig they were only related to the admin panel resource table.

**Reply by danharrin** [reply]
The admin panel uses the table builder so any features are mirrored

**Reply by maccarib** [reply]
Yes, but it isn't necessarily the other way round, right? I'm sorry when I'm not deeply into the filament structure, I'm relatively new to that (butI love it already :-) ). 

But as far as I understand, you have the getDefaultTableSortColumn-method available in the Table Builder and accordingly in the admin panel ressource table. As of version 2 this function only supports on column.

In the admin resource table you have the defaultSort-method that can be already chained several times and therefore already supports several columns. But that function is not available in the Table Builder when you use the Table Builder outside of the admin panel.

My suggestion was to implement a solution that also supports more than one column in the Table Builder (used outside the admin panel).  Is that what you've implemented in v3? Or are we talking past each other?

**Reply by danharrin** [reply]
`defaultSort()` in the admin panel just sets the `getDefaultTableSortColumn()` method on the list page.

All table builder methods are customisable in the admin panel by adding them to the list page.

In v3, we have consolidated the API so that you can configure tables in the table builder identically to how you configure tables in the admin panel, with no list page overrides.

**Reply by maccarib** [reply]
Yes, I've found that in the code.
But I think I've now also found the root of my misunderstanding. I somehow came to a theory based on information from the internet that I can't find again ;-), that in the admin panel resource table in v2 you can already do something like for example
`->defaultSort('name')->defaultSort('date')`
Which is simply not true, when I get it right, but is the key to all I've written above. I'm sorry for the extra rounds, and thank you very much for the detailed replies.
Anyways, if that feature comes with v3 that's great and will be enough for now.

**Reply by incloudsistemas** [reply]
->defaultSort(function (Builder $query): Builder {
                return $query
                    ->orderBy('name')
                    ->orderBy('date');
            })

**Reply by Nestoran** [reply]
> ->defaultSort(function (Builder $query): Builder { return $query ->orderBy('name') ->orderBy('date'); })

```Filament\Resources\Table::defaultSort(): Argument #1 ($column) must be of type string, Closure given...```


I feel like this should work on the ListPage, but it isn't either:
```php
protected function applyDefaultSortingToTableQuery(Builder $query): Builder
    {
        return $query
            ->orderBy('name')
            ->orderBy('date');
    }
```

**Reply by maccarib** [reply]
I'm still not so sure if we're talking about the same situation. Yes, in v3 you have `->defaultSort('stock', 'desc');` available at a Filament-Table.
But in the context of a livewire component that `InteractsWithTable`, you don't have access to that function. There's only the `protected function getDefaultTableSortColumn(): ?string` available, which supports only one column, and `InteractsWithTable` simply calls `->defaultSort($this->getDefaultTableSortColumn(), $this->getDefaultTableSortDirection())`.

In my initial post I was asking for something like `getDefaultTableSortColumns` in a livewire component/at `InteractsWithTable`,

With an orderBy-approach I can achieve a sorting by 2 or more columns, but I loose the ability to use an interactive sorting by the user through the table headers.

**Reply by danharrin** [reply]
`InteractsWithTable` now also has `defaultSort()` since the introduction of the fluent `table()` method in v3

**Reply by maccarib** [reply]
OK, great! I didn't realize that, I'm sorry! That works now! Thanks for the reply!
